db/repos/accounts.js

```diff
@@ -25,34 +30,414 @@ var columnSet;
+			{name: 'u_multisignatures', mod: ':raw', init: () => '(SELECT ARRAY_AGG("dependentId") FROM mem_accounts2u_multisignatures WHERE "accountId" = "mem_accounts"."address")'}
+		];
+
+		this.dbFields = _.union(normalFields, immutableFields, dynamicFields);
+
+		if (!columnSet) {
+			columnSet = {};
+
+			const table = new pgp.helpers.TableName({table: this.dbTable, schema: 'public'});
+
+			columnSet.select = new pgp.helpers.ColumnSet(this.dbFields, {table: table});
+
+			columnSet.update = new pgp.helpers.ColumnSet(normalFields, {table: table});
+			columnSet.update = columnSet.update.merge([
+				{name: 'publicKey', mod: ':raw', init: (object) => (object.value === undefined || object.value === null) ? 'NULL' : `DECODE('${object.value}', 'hex')`, skip: ifNotExists},
+				{name: 'secondPublicKey', mod: ':raw', init: (object) => (object.value === undefined || object.value === null) ? 'NULL' : `DECODE('${object.value}', 'hex')`, skip: ifNotExists},
+				{name: 'isDelegate', cast: 'int', def: 0, skip: ifNotExists},
+				{name: 'u_isDelegate', cast: 'int', def: 0, skip: ifNotExists},
+				{name: 'secondSignature', cast: 'int', def: 0, skip: ifNotExists},
+				{name: 'u_secondSignature', cast: 'int', def: 0, skip: ifNotExists},
+				{name: 'virgin', cast: 'int', def: 1},
+			]);
+
+			columnSet.insert = columnSet.update.merge([
+				{name: 'address', mod: ':raw', init: (object) => `UPPER('${object.value}')`}
+			]);
+		}
+
+		this.cs = columnSet;
+	}
+
+	/**
+	 * Get list of all database fields
+	 *
+	 * @return {array}
+	 */
+	getDBFields () {
+		return _.map(this.dbFields, (field) => (field.prop || field.name));
+	}
+
+	getImmutableFields () {
+		return _.difference(
+			_.map(this.cs.insert.columns, (field) => (field.prop || field.name)),
+			_.map(this.cs.update.columns, (field) => (field.prop || field.name)));
+	};
+
+	/**
+	 * Count mem accounts
+	 * @return {Promise}
+	 */
+	countMemAccounts () {
+		return this.db.one(Queries.countMemAccounts);
+	}
+
+	/**
+	 * Update mem accounts
+	 * @return {Promise}
+	 */
+	updateMemAccounts () {
+		return this.db.none(Queries.updateMemAccounts);
+	}
+
+	/**
+	 * Get orphan mem accounts
+	 * @return {Promise}
+	 */
+	getOrphanedMemAccounts () {
+		return this.db.query(Queries.getOrphanedMemAccounts);
+	}
+
+	/**
+	 * Get delegates
+	 * @return {Promise}
+	 */
+	getDelegates () {
+		return this.db.query(Queries.getDelegates);
+	}
+
+	/**
+	 * Update or insert into mem_accounts
+	 * @param {Object} data - Attributes to be inserted, can be any of [AccountsRepo's dbFields property]{@link AccountsRepo#cs.insert}
+	 * @param {Array} conflictingFields - Array of attributes to be tested against conflicts, can be any of [AccountsRepo's dbFields property]{@link AccountsRepo#dbFields}
+	 * @param {Object} updateData - Attributes to be updated, can be any of [AccountsRepo's dbFields property]{@link AccountsRepo#cs.update}
+	 * @return {Promise}
+	 */
+	upsert (data, conflictingFields, updateData) {
```

Comment:
For me thighs like upsert (`INSERT ... ON CONFLICT DO UPDATE`) are not acceptable in our system, from few reasons:
- performance, there is far less inserts than updates on `mem_accounts` table
- reliability - we should always know if there is a need to insert or update, and not have hope for best

Reply:
@4miners In first phase I simply migrated from `json-sql` to `pg-promise`. We already using this `INSERT.... ON CONFLICT DO UPDATE`...

https://github.com/LiskHQ/lisk/blob/e6ef692a0803891e546e7d00eef0f7c93c8323cd/logic/account.js#L840-L846

I agree to you that we should not use this... But the question is do we refactor this in this PR or inn other improvement cycle.

Reply:
@karmacoma I had discussed with @4miners . You can decide should we do this change in this PR or later in seperate.

Reply:
Ok